### DISCUSSION TITLE
[Feature Request] Import Channels, Roles, Permission from Discord To Revolt

### DISCUSSION THREAD

**Discussion by VulpoTheDev** [stoatchat/discussions#358] [Feature Suggestions]
### What do you want to see?

Migrating a Discord Server all over to Revolt is going to be a pain. You have to get the roles, channels, permission all of that stuff. I was thinking about a way where users can import all of their discord server stuff onto their new/existing revolt server. 

## How we could Tackle this

- Users can invite a Bot to their discord server in which they can do a command and the bot will gather all the guild data and put it onto a database (Which will expire in 48 hours). Once completed it'll provide the UID which can used to refer to that Document in the Database
- Have an option when you press "Add Server" where you have an input field where you can paste the UID for that Document from the database and it'll go through the process of making the server with the data from the Document.

### PWA

- [ ] Yes, this feature request is specific to the PWA.

**Comment by Rexogamer** [comment]
I can't currently transfer issues to revoltchat/revolt and as such will leave this open here for the time being, but this feels better suited as a meta-task/discussion there - this would probably require quite a bit of backend work (in fact, the Revite side of things would be fairly light work - just a new option in the Add Server modal), and there are several ways this could be implemented so a discussion on the best way forward would be a good idea

**Reply by treminaor** [reply]
Is there any update on this? My community is seriously considering a switch over to Revolt due to discords policies and it would really be helpful if we could import our roles (and chat history would be a bonus)

**Reply by VulpoTheDev** [reply]
I could see if I can implement make some sort of bot maybe from discord <-> Revolt

**Reply by treminaor** [reply]
It took about 3-4 hours but I did manage to replicate our server and permissions structure including role granting from the Roles bot. 

Now the main concern is chat history. Is there any feasible way to import that from Discord? Will that ever be possible?

**Reply by aScHEABL** [reply]
@treminaor It's definitely feasible, if you want it now you just have to develop a bot or write script on browser yourself.
Pulling data with 30+ text channels with over 4+ years of activities, one would require a 10~30 terabytes of storage space.
And the process likely will lasts a few days to a few weeks depending on how powerful your PC and how much chat history there is.

**Reply by Natans8** [reply]
A thing we should also remember, that Discord API is rate limited by intent. And you can only download a channel history by going over the messages one by one. The current rate is 50 requests per second, which is really really slow when you need to export large channels.

**Reply by IAmTomahawkx** [reply]
message history won't happen. Scraping channel history is a no-no from discord's pov, and also raises privacy concerns for user data (server owners do not have the right to other users messages).

**Reply by Puhpine4915** [reply]
I think such a bot/script would still be very valuable even without chat history. But from your messages I assume there is no bot/script like this yet?

**Reply by aScHEABL** [reply]
> message history won't happen. Scraping channel history is a no-no from discord's pov

I'm sorry to be blunt but... who the f*** cares about discord TOS if I'm leaving discord in the first place.
They were the first to make the ass*** move towards their existing users.
Even if you requested a complete removal of your data on discord, the only thing they do is to make your username "Discord user ####" instead of deleting the actual messages you sent. What's even the point?

I think discord users who want to migrate should just take matters into own hands.
@Puhpine4915 nope, not yet. I haven't finished reading docs on both platforms. And I'm far too busy in real life...

**Reply by IAmTomahawkx** [reply]
> I'm sorry to be blunt but... who the f*** cares about discord TOS if I'm leaving discord in the first place.

We who do not wish to be sued. We are an entity that Discord could choose to prosecute legally, since this is a flagrant violation of their terms of service in order to better a competing platform.
